File: samples/Kawa.Sample.VB.UseCases/CreateUser.vb
Imports System
Imports System.Threading
Imports System.Threading.Tasks
Imports Kawa.Abstractions

Namespace Kawa.Sample.VB.UseCases
    ''' <summary>Carries the name for the VB.NET sample create-user request.</summary>
    Public NotInheritable Class CreateUserRequest
        ''' <summary>Gets or sets the user name to create.</summary>
        Public Property Name As String = ""
    End Class

    ''' <summary>Carries the VB.NET sample create-user response.</summary>
    Public NotInheritable Class CreateUserResponse
        ''' <summary>Initializes a new instance of the <see cref="CreateUserResponse" /> class.</summary>
        Public Sub New()
        End Sub

        ''' <summary>Initializes a new instance of the <see cref="CreateUserResponse" /> class.</summary>
        ''' <param name="id">The generated sample user identifier.</param>
        ''' <param name="name">The created user name.</param>
        Public Sub New(id As String, name As String)
            Me.Id = id
            Me.Name = name
        End Sub

        ''' <summary>Gets or sets the generated sample user identifier.</summary>
        Public Property Id As String = ""

        ''' <summary>Gets or sets the created user name.</summary>
        Public Property Name As String = ""
    End Class

    ''' <summary>Creates a simple user response from the VB.NET sample use case.</summary>
    Public NotInheritable Class CreateUserUseCase
        Implements IUseCase(Of CreateUserRequest, CreateUserResponse)

        ''' <inheritdoc />
        Public Function ExecuteAsync(
            request As CreateUserRequest,
            Optional cancellationToken As CancellationToken = Nothing) As Task(Of KawaResult(Of CreateUserResponse)) _
            Implements IUseCase(Of CreateUserRequest, CreateUserResponse).ExecuteAsync

            cancellationToken.ThrowIfCancellationRequested()

            Dim response = New CreateUserResponse(Guid.NewGuid().ToString("N"), request.Name)
            Return Task.FromResult(KawaResult(Of CreateUserResponse).Success(response))
        End Function
    End Class
End Namespace

File: samples/Kawa.Sample.FSharp.UseCases/CreateUser.fs
namespace Kawa.Sample.FSharp.UseCases

open System
open System.Threading
open System.Threading.Tasks
open Kawa.Abstractions
open Kawa.FSharp

[<AllowNullLiteral>]
/// <summary>Carries the name for the F# sample create-user request.</summary>
type CreateUserRequest() =
    /// <summary>Gets or sets the user name to create.</summary>
    member val Name = "" with get, set

/// <summary>Carries the F# sample create-user response.</summary>
/// <param name="id">The generated sample user identifier.</param>
/// <param name="name">The created user name.</param>
type CreateUserResponse(id: string, name: string) =
    /// <summary>Gets the generated sample user identifier.</summary>
    member _.Id = id

    /// <summary>Gets the created user name.</summary>
    member _.Name = name

/// <summary>Creates a simple user response from the F# sample use case.</summary>
type CreateUserUseCase() =
    interface IUseCase<CreateUserRequest, CreateUserResponse> with
        /// <inheritdoc />
        member _.ExecuteAsync(request: CreateUserRequest, cancellationToken: CancellationToken) =
            cancellationToken.ThrowIfCancellationRequested()

            CreateUserResponse(Guid.NewGuid().ToString("N"), request.Name)
            |> KawaResult.success
            |> Task.FromResult
